text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,392 @@
+Instead of creating a new binding, you can also assign to an existing one by leaving out the `let`, similarly to destructuring assignment:
+
+```rust
+fn example(x: Foo) {
+    let mut a = vec![1, 2, 3];
+    
+    destructure!(Foo { a, b: _ } = x);
+    // yay we own `a` now, we can even drop it!
+    drop(a);
+}
+```
+
+When a type implements `Drop`, this can be because the order in which fields are dropped is very important. It could be unsound to do it in the wrong order. When you write a destructuring, soundly dropping the fields is now your responsibility.
+
+This can be a heavy burden, so if you are the author of a module or crate, you might want to limit other people destructuring your type. The rule for this is that you can only use `destructure!()` on types in a location where you could also construct that type. This means that in that location, all fields must be visible.
+
+Importantly, this means that this code is not valid:
+
+```rust
+mod foo {
+    pub struct Bar {
+        pub a: Vec<u8>,
+        b: Vec<u8>,
+    }
+}
+
+fn example(x: foo::Bar) {
+    destructure!(let Foo{ a, .. } = x)
+}
+```
+
+By using `..`, we could ignore the fact that `b` was not visible and move out `a` anyway. This is undesirable, as it would implicitly run `drop(b)` even though `b` was not accessible here. In fact, if there was a requirement that `a` was dropped before `b`, it would never be sound to destructure a type `Bar` in a location where `b` is not accessible.
+
+Finally, you might ask why we need a macro for this at all. This is because for some types, the behavior is slightly different. 
+Let's use the following example. In Rust right now, this does not compile:
+
+```rust
+struct Foo {
+    a: Vec<u64>
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // error: cannot move out of type `Foo`, which implements the `Drop` trait
+    let Foo { a } = x;
+}
+```
+
+This is because we move the field `a` out, and `Foo` implements `Drop`.
+The following does compile:
+
+```rust
+struct Foo {
+    a: Vec<u64>
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // fine, because we don't actually move any fields out
+    let Foo { .. } = x;
+}
+```
+
+Similarly, the following also works because we don't have to move out any fields, we can just copy `a`, because `u64` implements `Copy`.
+
+```rust
+struct Foo {
+    a: u64
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // Totally fine, we just copy a
+    let Foo { a } = x;
+}
+```
+
+In the two examples above, `Drop` still runs for `x`.
+`destructure!()` represents a different operation. When we use it, we actually move the fields out of the type, and prevent `Drop` from running.
+
+```rust
+struct Foo {
+    a: u64
+}
+
+impl Drop for Foo {
+    fn drop(&mut self) {}
+}
+
+fn example(x: Foo) {
+    // `a` is moved out of `x`, Drop never runs for `x`
+    destructure!(let Foo { a } = x);
+}
```

Comment:
I've definitely written code that would have been much-improved... or at least written faster... by this being available! Thank you!